### PR TITLE
[libc++][Android] Support libc++ testing on Android

### LINKED ISSUE #69270: [libc++][Android] Add official Android support to libc++
Android switched to libc++ years ago, but Android still isn't officially supported in libc++. We should add support for it. libc++ already builds and works on Android, but many of the tests don't pass, and there should be automated testing.

### PR DESCRIPTION
**Pull request by rprichard**
I could probably break this commit into more pieces.

---

This patch adds libc++ support for Android L (Android 5.0+) and up, tested using the Android team's current compiler, a recent version of the AOSP sysroot, and the x86[-64] Android Emulator.

CMake and Lit Configuration:

Add runtimes/cmake/android/Arch-${ARCH}.cmake files that configure CMake to cross-compile to Android without using CMake's built-in NDK support (which only works with an actual packaged NDK).

Add libcxx/cmake/caches/AndroidNDK.cmake that builds and tests libc++ (and libc++abi) for Android. This file configures libc++ to match what the NDK distributes, e.g.:
 - libc++_shared.so (includes libc++abi objects, there is no libc++abi.so). libunwind is linked statically but not exported.
 - libc++_static.a (does not include libc++abi) and libc++abi.a
 - `std::__ndk1` namespace
 - All the libraries are built with `__ANDROID_API__=21`, even when they are linked to something targeting a higher API level.

(However, when the Android LLVM team builds these components, they do not use these CMake cache files. Instead they use Python scripts to configure the builds. See
https://android.googlesource.com/toolchain/llvm_android/.)

Add llvm-libc++[abi].android-ndk.cfg.in files that test the Android NDK's libc++_shared.so. These files can target old or new Android devices. The Android LLVM team uses these test files to test libc++ for both arm/arm64 and x86/x86_64 architectures.

The Android testing mode works by setting %{executor} to adb_run.py, which uses `adb push` and `adb shell` to run tests remotely. adb_run.py always runs tests as the "shell" user even on an old emulator where "adb unroot" doesn't work. The script has workarounds for old Android devices. The script uses a Unix domain socket on the host (--job-limit-socket) to restrict concurrent adb invocations. Compiling the tests is a major part of libc++ testing run-time, so it's desirable to exploit all the host cores without overburdening the test devices, which can have far fewer cores.

BuildKite CI:

Add a builder to run-buildbot, `android-ndk-*`, that uses Android Clang and an Android sysroot to build libc++, then starts an Android emulator container to run tests.

Run the emulator and an adb server in a separate Docker container (libcxx-ci-android-emulator), and create a separate Docker image for each emulator OS system image. Set ADB_SERVER_SOCKET to connect to the container's adb server. Running the only adb server inside the container makes cleanup more reliable between test runs, e.g. the adb client doesn't create a `~/.android` directory and the adb server can be restarted along with the emulator using docker stop/run. (N.B. The emulator insists on connecting to an adb server and will start one itself if it can't connect to one.)

The suffix to the android-ndk-* job is a label that concisely specifies an Android SDK emulator image. e.g.:
 - "system-images;android-21;default;x86" ==> 21-def-x86
 - "system-images;android-33;google_apis;x86_64" ==> 33-goog-x86_64

Fixes: https://github.com/llvm/llvm-project/issues/69270
Differential Revision: https://reviews.llvm.org/D139147

### CODE REVIEW

File: libcxx/docs/index.rst
Comment (EricWF) [inline libcxx/docs/index.rst:1]: I'm fine marking android as supported now, but others may chime in later.

**Review by EricWF — APPROVED** [review]
This looks good to me. There's a lot of code here, but very little of it modifies anything that is existing.

Let's land this and see how it goes.

File: libcxx/docs/index.rst
Comment (ldionne) [reply libcxx/docs/index.rst:1]: To mark it as supported, we need to have a BuiltKite runner enabled. @rprichard how far along are you with setting that up?

Comment (rprichard) [reply libcxx/docs/index.rst:1]: It's getting close? I think @EricWF is looking at it at the moment?

(I could remove this line for now.)

Comment (rprichard) [reply libcxx/docs/index.rst:1]: I suppose I should have put that change in the final commit enabling the BuildKite CI...

Comment (rprichard) [reply libcxx/docs/index.rst:1]: I uploaded https://github.com/llvm/llvm-project/pull/69660 reverting that change to the doc.

**Comment by rprichard** [comment]
It looks like black/darker rejects the new Python code I added, but somehow the GitHub check passed anyway (https://github.com/llvm/llvm-project/pull/69274/checks).

After looking at llvm/utils/git/code-format-helper.py some more, I see that it is discarding stderr output from `darker` and from `git-clang-format`.

https://github.com/llvm/llvm-project/blob/7f7a15c3579c4d736e1cfaa7a4097f5a7afd9526/llvm/utils/git/code-format-helper.py#L145-L151

https://github.com/llvm/llvm-project/blob/7f7a15c3579c4d736e1cfaa7a4097f5a7afd9526/llvm/utils/git/code-format-helper.py#L94-L101

`capture_output=True` means `stdout=PIPE stderr=PIPE`, but `proc.stderr` is discarded. `""` and `None` are both false, so if the code-formatter fails to run, that's considered a quiet success. It should at least record the stderr somewhere, and probably should fail the check, so someone notices.

**Pull request merged**